cdap-app-templates/cdap-etl/cdap-etl-lib/src/main/java/co/cask/cdap/template/etl/transform/LogParserTransform.java

```diff
@@ -0,0 +1,203 @@
+  //Number of groups matched in the S3 regex
+  private static final int S3_REGEX_LENGTH = 18;
+  //Number of groups matched in the CLF regex
+  private static final int CLF_REGEX_LENGTH = 9;
+  private static final String S3_LOG = "S3";
+  private static final String CLF_LOG = "CLF";
+  private final LogParserConfig config;
+
+  public LogParserTransform(LogParserConfig config) {
+    this.config = config;
+  }
+
+  @Override
+  public void transform(StructuredRecord input, Emitter<StructuredRecord> emitter) throws Exception {
+    String log;
+    StructuredRecord output;
+    Schema.Field inputField = input.getSchema().getField(config.inputName);
+    if (inputField == null) {
+      LOG.error("Invalid inputName, no known field matches given input.");
+      return;
+    }
+    Schema.Type inputType = inputField.getSchema().getType();
+    if (Schema.Type.STRING.equals(inputType)) {
+      log = input.get(config.inputName);
+    } else if (Schema.Type.BYTES.equals(inputType)) {
+      Object data = input.get(config.inputName);
+      if (data instanceof ByteBuffer) {
+        log = Bytes.toString((ByteBuffer) input.get(config.inputName));
+      } else if (data instanceof byte[]) {
+        log = Bytes.toString((byte[]) input.get(config.inputName));
+      } else {
+        LOG.debug("Not a byte type");
+        return;
+      }
+    } else {
+      LOG.error("Unsupported inputType, only bytes and string are supported.");
+      return;
+    }
+
+    if (log == null) {
+      LOG.error("Couldn't read schema, log message was null");
+      return;
+    }
+
+    if (S3_LOG.equals(config.logFormat)) {
+      Matcher logMatcher = S3_LOG_PATTERN.matcher(log);
+      if (!logMatcher.matches() || logMatcher.groupCount() < S3_REGEX_LENGTH) {
+        LOG.debug("Couldn't parse log because log did not match the S3 format, log:" + log);
+        return;
+      }
+      output = parseRequest(logMatcher, S3_INDICES);
+    } else if (CLF_LOG.equals(config.logFormat)) {
+      Matcher logMatcher = CLF_LOG_PATTERN.matcher(log);
+      if (!logMatcher.matches() || logMatcher.groupCount() < CLF_REGEX_LENGTH) {
+        LOG.debug("Couldn't parse log because the log did not match the CLF format. log:" + log);
+        return;
+      }
+      output = parseRequest(logMatcher, CLF_INDICES);
+    } else {
+      LOG.error("Log format not currently supported.");
+      return;
+    }
+    if (output != null) {
+      emitter.emit(output);
+    }
+  }
+
+  /**
+   * Parses a request for the URI, IP, Browser, Device, and Time
+   * @param logMatcher the regex matcher to use
+   * @param indices array of indices that define what position in the regex the fields are,
+   *                 in the order of Request, Time, IP, and User Agent
+   */
+  private StructuredRecord parseRequest(Matcher logMatcher, int[] indices) {
+    String request = logMatcher.group(indices[0]);
+    long ts = System.currentTimeMillis();
```

Comment:
I think it makes sense to have this at least 6 lines down, closer to the code block where it's used.